phalcon/acl/adapter/memory.zep

```diff
@@ -678,78 +679,94 @@ class Memory extends Adapter
-						}
-						// Number of required parameters == 0 so call funcAccess without any arguments
-						return (haveAccess == Acl::ALLOW) && call_user_func(funcAccess);
+
+			// No parameters, just return haveAccess and call function without array
+			if parameterNumber === 0 {
+				return haveAccess == Acl::ALLOW && call_user_func(funcAccess);
+			}
+
+			let parametersForFunction = [];
+			let numberOfRequiredParameters = reflectionFunction->getNumberOfRequiredParameters();
+			let userParametersSizeShouldBe = parameterNumber;
+
+			for reflectionParameter in reflectionParameters {
+				let reflectionClass = reflectionParameter->getClass();
+				let parameterToCheck = reflectionParameter->getName();
+
+				if reflectionClass !== null {
+					// roleObject is this class
+					if roleObject !== null && reflectionClass->isInstance(roleObject) {
+						let parametersForFunction[] = roleObject;
+						let userParametersSizeShouldBe--;
+
+						continue;
 					}
-					// Check necessary parameters
-					elseif count(parametersForFunction) >= numberOfRequiredParameters {
-						return (haveAccess == Acl::ALLOW) && call_user_func_array(funcAccess,parametersForFunction);
+
+					// resourceObject is this class
+					if resourceObject !== null && reflectionClass->isInstance(resourceObject) {
+						let parametersForFunction[] = resourceObject;
+						let userParametersSizeShouldBe--;
+
+						continue;
 					}
-					// We dont have enough parameters
-					else {
-						throw new Exception("You didn't provide all necessary parameters for defined function when check ".roleName." can ".access." ".resourceName);
+
+					// This is some user defined class, check if his parameter is instance of it
+					if isset parameters[parameterToCheck] && typeof parameters[parameterToCheck] == "object" && !reflectionClass->isInstance(parameters[parameterToCheck]) {
+						throw new Exception(
+							"Your passed parameter doesn't have the same class as the parameter in defined function when check " . roleName . " can " . access . " " . resourceName . ". Class passed: " . get_class(parameters[parameterToCheck])." , Class in defined function: " . reflectionClass->getName() . "."
+						);
 					}
+				}
+
+				if isset parameters[parameterToCheck] {
+					// We can't check type of ReflectionParameter in PHP 5.x so we just add it as it is
+					let parametersForFunction[] = parameters[parameterToCheck];
+				}
+			}
+
+			if count(parameters) > userParametersSizeShouldBe {
+				trigger_error(
+					"Number of parameters in array is higher than the number of parameters in defined function when check " . roleName . " can " . access . " " . resourceName . ". Remember that more parameters than defined in function will be ignored.",
+					E_USER_WARNING
+				);
 			}
+
+			// We dont have any parameters so check default action
+			if count(parametersForFunction) == 0 {
+				if numberOfRequiredParameters > 0 {
+					trigger_error(
```

Comment:
Well when i was creating it i added this, but should we inform developer that we will use this check ? Or maybe remove this trigger_error ? @sergeyklay